### PR TITLE
Fix camera to follow player eyes when player is attached 

### PR DESCRIPTION
**Pull request by sfence**
- Goal of the PR
Make the player camera follow the player's eyes independently on attached offset and rotation and parent object rotation.
- How does the PR work?
Change the camera position setting.
- Does it resolve any reported issue?
This probably resolves bug issue #10101 and more.
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
I have no idea.

## To do

This PR is a Ready to Review

## How to test

Run devtest game, add `testentities:attach` entity to the scene, and right-click to attach yourself to it.
Control keys are up, down, left, right, jump, sneak, and aux1. `aux1` changes the mode of control.
To check `get_look_dir`, take the `testlookdir:fire` item and use the dig/place buttons to the fire entity in the direction of `look_dir` get from `get_look_dir` or calculated from `get_look_vertical` and `get_look_horizontal`.

You can detach this branch HEAD to one commit before commit "Fix camera position and rotation." to see how it works without a camera fix.

See before :
https://streamable.com/ih3lb5

And after:
https://streamable.com/tvdztt

### CODE REVIEW

**Comment by sfence** [comment]
@APercy This can be interesting for you.

File: src/client/camera.cpp [hunk 1]
```diff
@@ -316,16 +316,29 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 tool_reload_ratio)
...
+	if (player->getParent()) {
+		GenericCAO *parent = dynamic_cast<GenericCAO *>(player->getParent());
+		int parent_id;
+		v3f attach_pos;
+		v3f attach_rot;
+		std::string bone;
+		bool force_visible;
+		GenericCAO *player_cao = player->getCAO();
+		player_cao->getAttachment(&parent_id, &bone, &attach_pos, &attach_rot, &force_visible);
+		parent_rotation = parent->getRotation() - attach_rot;
```
Comment (appgurueu) [inline src/client/camera.cpp:337]: I don't think this is correct in general. You want to ~~undo~~ apply `attach_rot` here, but you can't do that by subtracting Euler angles. You'd have to use quaternions: First convert to quaternions, then invert the quaternion for the attachment rotation (different conventions for rotation orientation?), then multiply that with the quaternion for the parent rotation, then convert back to Euler angles (and be careful about radians vs degrees).

**Review by appgurueu — COMMENTED** [review]
Such a fix is much needed and useful, but I have some concerns about breakages, esp. since there are no changes to the Lua API.

Before this PR, the camera did not have roll, and even though the player visuals could be rotated, the eye position could always be found by adding eye height to the player pos. This PR changes that, and gives mods no convenient way to detect this or to find the new eye pos; mods would have to query attachments and compose transformations themselves.
(I'm unfortunately not sure whether we can expose a good server-side API, since the player might be attached to an animated model, which the server doesn't understand. But at least for the common special case where no bones are involved, we could make a decent effort.)

Calculating look dir from pitch and yaw will also be broken (there is no way around this when adding roll, but again, modders need to be able to deal with it.) Also, won't `get_look_dir` be broken (as in: was absolute, now is relative to attachment) too?

What it boils down to: Wouldn't this break firing a bow from the test entity, both in terms of finding the eye position and the direction the arrow should go? And if so, how can we best update the API such that modders can deal with this?

File: src/client/camera.cpp (hunk 1 above)
Comment (sfence) [reply src/client/camera.cpp:337]: I tried to do this:

```
     player_cao->getAttachment(&parent_id, &bone, &attach_pos, &attach_rot, &force_visible);
     core::quaternion quat_rot(parent->getRotation() * core::DEGTORAD);
     quat_rot = quat_rot * core::quaternion(-attach_rot * core::DEGTORAD);
     parent_rotation = parent->getRotation() - attach_rot; // not used here
     quat_rot.toEuler(parent_rotation);
     parent_rotation = parent_rotation * core::RADTODEG;
     attach_pos += player->getEyeOffset();
     parent->getPosRotMatrix().rotateVect(attach_pos);
     player_position = parent->getPosition() + attach_pos;
```

And it stops the camera from following the player's eyes. It follows only parent node rotation and attach_rot makes no difference. I tried also `quat_rot = core::quaternion(-attach_rot * core::DEGTORAD) * quat_rot;`, but without visible effect.

I have debugged it with vectors like:

attach_rot = {X = -5.26318979, Y = 0, Z = 93.7392807}
parent_rotation = {X = 355.990173, Y = 0, Z = 0.400718689} -> after sub with attach_rot

and for this, quaternion gets the result {X = -9.65404415, Y = -5.24941778, Z = 0.41818282} in DEG

Based on https://stackoverflow.com/questions/56246893/how-do-i-rotate-a-quaternion-with-2nd-quaternion-on-its-local-or-world-axes-with

Another try:

```
     player_cao->getAttachment(&parent_id, &bone, &attach_pos, &attach_rot, &force_visible);
     parent_rotation = parent->getRotation() - attach_rot;
     core::quaternion quat_rot(parent_rotation * core::DEGTORAD);
     quat_rot.toEuler(parent_rotation);
     parent_rotation = parent_rotation * core::RADTODEG;
     attach_pos += player->getEyeOffset();
     parent->getPosRotMatrix().rotateVect(attach_pos);
```

Both with/without:
```
if(parent_rotation.X < 0)
    parent_rotation.X += 360;
...
```

And some results as the previous try. Camera positions follow only parent rotation, and `attach_rot` makes no visible effect.

Based on https://irrlicht.sourceforge.io/forum/viewtopic.php?p=285337

Comment (sfence) [reply src/client/camera.cpp:337]: Finnaly this looks to works better:

```
     core::quaternion quat_rot(parent->getRotation() * core::DEGTORAD);
     quat_rot = quat_rot * core::quaternion(-attach_rot * core::DEGTORAD);
     parent_rotation = parent->getRotation() - attach_rot; // not used here
     quat_rot.toEuler(parent_rotation);
     parent_rotation = parent_rotation * core::RADTODEG;
     quat_rot.normalize();
     v3f eye_offset = player->getEyeOffset();
     quat_rot.getMatrix().inverseRotateVect(eye_offset);
     attach_pos += eye_offset;
     //parent->getPosRotMatrix().rotateVect(attach_pos);
     
     player_position = parent->getPosition() + attach_pos;

```

But camera rotation stops following parent and attachment rotation.

Comment (sfence) [reply src/client/camera.cpp:337]: Please, check last commit.
This solutions looks to be working well.

i still have to update pich and yawto be relative too and similar. But that code will be delivered from from camera orientation code, I expect.

**Comment by sfence** [comment]
Works pretty well now until attachment rotation uses the roll (axis Z in vector) part of the rotation.
`get_look_vertical` and `get_look_horizontal` still unchanged.

**Comment by sfence** [comment]
So, now the matrixes are used to calculate positions and rotation in the way, how it Is done on the client side.
World relative pitch and yaw are calculated from the camera direction vector on the client side and sent to the server.
With this, pitch and yaw should work naturally on old and new clients on the same server.
Position returned by Lua `minetest.get_pos` respect attachments and rotations.

**Comment by sfence** [comment]
@appgurueu Is it a good idea to detect the player client protocol version in `minetest.get_eye_pos` and `minetest.get_pos` and return values in the old style for players playing on old clients?

**Comment by Zughy** [comment]
@sfence rebase needed

**Pull request closed without merging**